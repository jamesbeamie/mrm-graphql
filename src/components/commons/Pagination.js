import React, { Component } from 'react';
import PropTypes from 'prop-types';
import '../../assets/styles/pagination.scss';

const options = array =>
  array.map(item => <option key={item + 1}>{item}</option>);

/**
 * Reusable pagination component
 *
 * @extends Component
 *
 * @returns {JSX}
 */
class Pagination extends Component {
  state = {
    perPage: 5,
    page: 1,
  };

  componentDidMount = () => {
    const { currentPage } = this.props;
    this.setState({ page: currentPage });
  }

  /**
   * Handles changes to the pagination field
   *
   * @returns {void}
   */
  handleChange = (e) => {
    e.preventDefault();
    const { name, value } = e.target;
    const { perPage } = this.state;
    const { handleData, dataFetched } = this.props;
    if (name === 'perPage' && dataFetched) {
      this.setState({ perPage: parseFloat(value), page: 1 }, () => {
        handleData(value, 1);
      });
    }
    if (name === 'page' && dataFetched) {
      this.setState({ page: parseFloat(value), perPage }, () => {
        handleData(perPage, parseInt(value, 10));
      });
    }
  };

  /**
   * Calls data for the next page when the next button is hit.
   *
   * @returns {void}
   */
  handleNext = () => {
    const {
      handleData, dataFetched, perPage, currentPage,
    } = this.props;
    handleData(perPage, currentPage + 1);
    dataFetched &&
      this.setState({
        page: currentPage + 1,
      });
  };

  /**
   * Calls data for the previous page when the previous button is hit.
   *
   * @returns {Function} handleData
   */
  handlePrevious = () => {
    const {
      handleData, dataFetched, currentPage,
    } = this.props;
    const { perPage } = this.state;
    handleData(perPage, currentPage - 1);
    dataFetched &&
      this.setState({
        page: currentPage - 1,
      });
  };

  render() {
    const {
      itemsPerPage,
      totalPages,
      hasNext,
      hasPrevious,
      reverse,
      isFetching,
      perPage,
    } = this.props;
    const { page } = this.state;
    return (
      <nav className="page-navigation">
        <div className="perPageBlock">
          <ul className="pagination">
            <select
              name="perPage"
              value={perPage}
              id="perPage"
              className="page-select"
              onChange={this.handleChange}
              disabled={isFetching}
            >
              {options(itemsPerPage)}
            </select>
            <li>Items per page</li>
          </ul>
        </div>
        <div className="showingBlock">
          <ul className="pagination">
            <li>
              <button
                id="previous"
                onClick={this.handlePrevious}
                className={!isFetching && hasPrevious ? 'enabled' : 'disabled'}
              >
                Previous
              </button>
            </li>
            <li className="results">
              <ul>
                <span>Showing {reverse ? '' : 'page'}</span>
                <select
                  className="totalPage"
                  id="totalPage"
                  name="page"
                  value={page}
                  onChange={this.handleChange}
                  disabled={isFetching}
                >
                  {[...Array(totalPages).keys()].map(item => (
                    <option key={item}>{item + 1}</option>
                  ))}
                </select>
                {!reverse && <span className="pageNum"> of {totalPages}</span>}
                {reverse && (
                  <span className="pageNum reverse">
                    {' '}
                    of {totalPages} Results
                  </span>
                )}
              </ul>
            </li>
            <li>
              <button
                id="next"
                onClick={this.handleNext}
                className={hasNext && !isFetching ? 'enabled' : 'disabled'}
              >
                Next
              </button>
            </li>
          </ul>
        </div>
      </nav>
    );
  }
}

Pagination.propTypes = {
  perPage: PropTypes.number, // eslint-disable-line
  itemsPerPage: PropTypes.arrayOf(PropTypes.number),
  hasNext: PropTypes.bool,
  hasPrevious: PropTypes.bool,
  totalPages: PropTypes.number,
  handleData: PropTypes.func.isRequired,
  reverse: PropTypes.bool,
  currentPage: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  dataFetched: PropTypes.bool,
  isFetching: PropTypes.bool,
};

Pagination.defaultProps = {
  perPage: 5,
  itemsPerPage: [5, 10, 20, 50],
  totalPages: 4,
  currentPage: 1,
  hasNext: true,
  hasPrevious: false,
  reverse: false,
  dataFetched: true,
  isFetching: false,
};

export default Pagination;
