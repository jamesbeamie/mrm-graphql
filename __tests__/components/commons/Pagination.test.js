import React from 'react';
import { shallow } from 'enzyme';
import Pagination from '../../../src/components/commons/Pagination';

describe('Pagination Component', () => {
  const props = {
    totalPages: 4,
    hasNext: true,
    hasPrevious: false,
    perPage: 5,
    currentPage: 1,
    handleData: jest.fn(),
  };
  const paginationWrapper = shallow(<Pagination {...props} />);

  it('renders without crashing', () => {
    expect(paginationWrapper).toMatchSnapshot();
  });

  it('updates the page when a page input is provided', () => {
    const preventDefault = jest.fn();

    paginationWrapper.instance().handleChange({ preventDefault, target: { name: 'page', value: 3 } });
    expect(paginationWrapper.instance().state.page).toEqual(3);
  });

  it('updates perPage when a perPage input is provided', () => {
    const preventDefault = jest.fn();

    paginationWrapper.instance().handleChange({ preventDefault, target: { name: 'perPage', value: 10 } });
    expect(paginationWrapper.instance().state.perPage).toEqual(10);
  });

  it('navigates to the next page when next button is clicked', () => {
    paginationWrapper.find('#next').simulate('click');
    expect(paginationWrapper.instance().state.page).toEqual(2);
  });

  it('navigates to the previous page when previous button is clicked', () => {
    const nextProps = {
      totalPages: 4,
      hasNext: true,
      hasPrevious: true,
      perPage: 5,
      currentPage: 5,
      handleData: jest.fn(),
    };
    const prevWrapper = shallow(<Pagination {...nextProps} />);
    expect(prevWrapper.instance().state.page).toEqual(5);
    prevWrapper.find('#previous').simulate('click');
    expect(prevWrapper.instance().state.page).toEqual(4);
  });
  it('enables previous button if there is a next input', () => {
    const newProps = {
      totalPages: 4,
      hasNext: true,
      hasPrevious: true,
      handleData: jest.fn(),
      reverse: true,
    };
    const newPaginationWrapper = shallow(<Pagination {...newProps} />);
    expect(newPaginationWrapper.instance().props.hasPrevious).toEqual(true);
  });
  it('disables next button if there is no next input', () => {
    const newProps = {
      totalPages: 4,
      hasNext: false,
      hasPrevious: true,
      handleData: jest.fn(),
    };
    const newPaginationWrapper = shallow(<Pagination {...newProps} />);
    expect(newPaginationWrapper.instance().props.hasNext).toEqual(false);
  });


  it('disables paginator buttons on fetching data after click', () => {
    const newProps = {
      totalPages: 4,
      hasNext: true,
      hasPrevious: true,
      isFetching: true,
      handleData: jest.fn(),
    };

    const newPaginationWrapper = shallow(<Pagination {...newProps} />);
    expect(newPaginationWrapper.instance().props.isFetching).toEqual(true);
    expect(newPaginationWrapper.find('#next').hasClass('disabled')).toEqual(true);
    expect(newPaginationWrapper.find('#previous').hasClass('disabled')).toEqual(true);
    expect(newPaginationWrapper.find('#totalPage').is('[disabled]')).toBe(true);
    expect(newPaginationWrapper.find('#perPage').is('[disabled]')).toBe(true);
  });
});
