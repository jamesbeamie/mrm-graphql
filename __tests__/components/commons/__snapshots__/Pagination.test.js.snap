// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component renders without crashing 1`] = `
<nav
  className="page-navigation"
>
  <div
    className="perPageBlock"
  >
    <ul
      className="pagination"
    >
      <select
        className="page-select"
        disabled={false}
        id="perPage"
        name="perPage"
        onChange={[Function]}
        value={5}
      >
        <option
          key="6"
        >
          5
        </option>
        <option
          key="11"
        >
          10
        </option>
        <option
          key="21"
        >
          20
        </option>
        <option
          key="51"
        >
          50
        </option>
      </select>
      <li>
        Items per page
      </li>
    </ul>
  </div>
  <div
    className="showingBlock"
  >
    <ul
      className="pagination"
    >
      <li>
        <button
          className="disabled"
          id="previous"
          onClick={[Function]}
        >
          Previous
        </button>
      </li>
      <li
        className="results"
      >
        <ul>
          <span>
            Showing 
            page
          </span>
          <select
            className="totalPage"
            disabled={false}
            id="totalPage"
            name="page"
            onChange={[Function]}
            value={1}
          >
            <option
              key="0"
            >
              1
            </option>
            <option
              key="1"
            >
              2
            </option>
            <option
              key="2"
            >
              3
            </option>
            <option
              key="3"
            >
              4
            </option>
          </select>
          <span
            className="pageNum"
          >
             of 
            4
          </span>
        </ul>
      </li>
      <li>
        <button
          className="enabled"
          id="next"
          onClick={[Function]}
        >
          Next
        </button>
      </li>
    </ul>
  </div>
</nav>
`;
